### PR TITLE
Update python module guesser mappings

### PR DESCRIPTION
**Pull request by airportyh**
# Why?

We haven't updated these mappings in a couple of years, and they are out of date. The process is long running (worse estimated at 12-16 days) and is difficult and time consuming to do. The registry has grown a lot and the download numbers have also deviated a bit and some previous assumptions we made do not hold anymore.

## What Changed

0. Adds 4735 new packages to the guesser out of the top 10k
1. Incremental approach where the cache info for each package is saved one file per package in a `cache` directory, and is tagged by the package's version number.
2. gen_pypi_map.go now has 3 commands (modes of operation):
    1. `bq` - fetch download stats from pypi big table
    2. `test` - test modules from pypi, can be based on a set of packages passed in via `.json` file
    3. `gen` - generate the `pypi_map.gen.go` file
3. `bq.json` was renamed `download_stats.json`. Instead of merging download stats into the cache, they are kept separate.
3. Added a regression test file that compares the output of the previous version to this version for the 10k packages. Hand checked each deviation and either updated the algorithm or added manual override to the mapping or the test suite to account for them. Supporting files for regression tests (didn't not check in because they are large): https://gist.github.com/airportyh/5b56ad40528a64bcc3356bb0b05d863a
5. Added a pypi_map.override.go for manual overrides of the mapping when needed
6. Changed the heuristic algorithm for guessing to:
   1. check for exact or similar name matches: pick the most popular of the matches
   2. if 1 didn't work, filter out packages that depend on other packages on this list
   3. pick the most popular package of the remaining
   5. (still not perfect solution due to missing or incorrect info reported by packages)
7. `pkgs.json` is added as the source of truth for package dependency and module info for packages that have been tested, tagged by version. When running the install tests, info is saved into individual json files in the `cache` directory, but when finished, it will merge those results into `pkgs.json`. It was done this way so that if a job died, we can restart the process without losing work.
7. Ran install test for the top 10k modules based on download stats. Falling back to existing `pypi_packages.json` dataset.

## Test Plan

Use the regression test suite.

## Known Issues

1. Some of the module tests error out when installing the package due to missing native dependencies. We are falling back to the pypi_packages.json dataset where possible. There are 410 error cases in the top 10k currently. Plan to figure out native dependencies and create a knowledge base in the future.

### CODE REVIEW

**Review by dan-stowell — CHANGES_REQUESTED** [review]
What makes a good package mapping? You've told me that the package mapping you created using this code covers all but a few of the most popular 10,000 packages on PyPI. That's great! Can you commit a way to verify that?

If this code produces a good mapping, strictly better than our two-years-out-of-date mapping, I approve this code.

File: internal/backends/python/gen_pypi_map/install_diff.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	// Create a virtual enviornment to install the package to
+	cmd := exec.Command("python3", "-m", "venv", path)
+	err := cmd.Run()
+
+	if err != nil {
+		return nil, PypiError{InstallFailure, "Failed to create venv", err}
+	}
+
+	// Run a python script to find the newly installed modules
+	cmd = exec.Command("sh", "-c", ". "+path+"/bin/activate; python3 /home/toby/upm/internal/backends/python/gen_pypi_map/install_diff.py "+metadata.Info.Name)
```
[the remark this answers is not in the document]
Comment (airportyh) [reply internal/backends/python/gen_pypi_map/install_diff.go]: Thanks.

**Comment by airportyh** [comment]
@dan-stowell I have written up a README.md under the `internal/backends/python/regression_tests` folder with how to run the regression tests for yourself. The results I got:

5265 ok
4388 added
346 test errored

**Comment by airportyh** [comment]
> If this code produces a good mapping, strictly better than our two-years-out-of-date mapping, I approve this code.

@dan-stowell this is ready for you to take a look.

The regression tests take between 17-30 minutes to run, depending on hardware. You can see `internal/backends/python/regression_tests/README.md`. You can also spot check specific packages if you don't want to run the whole thing.

This is the output I got from my run: https://gist.github.com/airportyh/945fb5cf34a60cdcb38a3ce5e48e9274

This is the summary of those results:

```
toby@tobyval:~/upm/internal/backends/python/regression_tests$ cat result27.txt|grep ok|wc
   5265   10530   75416
toby@tobyval:~/upm/internal/backends/python/regression_tests$ cat result27.txt|grep added|wc
   2580    5160   52885
toby@tobyval:~/upm/internal/backends/python/regression_tests$ cat result27.txt|grep no-guess|wc
   1808    3616   53033
toby@tobyval:~/upm/internal/backends/python/regression_tests$ cat result27.txt|grep test-errored|wc
    348     696   11495
```

Possible improvements:
1. I've checked some of the no-guess ones and sometimes info gather in the testing phase isn't returning the correct module. I'll dig deeper into it.
2. For the test-errored ones, that's where we'll need figure out the native dependencies of the module

**Comment by dan-stowell** [comment]
> @dan-stowell this is ready for you to take a look.

Looking!

**Pull request merged**